lib/serialization/compositeSpec.ts

```diff
@@ -129,76 +201,188 @@ export function compositeSpec(typeName: string, propertySpecs: { [propertyName:
-            }
-          } else {
-            // JSON
-            serializedChildPropertyName = childPropertySpec.serializedName || childPropertyName;
-          }
 
-          // This part is for handling property flattening. If the serialized name has dots in its
-          // name (a.b.c), then we handle those name parts (a, b, c) as individual levels in the
-          // serialized value.
-          const serializedChildPropertyNameParts: string[] = splitSerializeName(serializedChildPropertyName);
-          if (serializedChildPropertyNameParts.length > 1) {
-            for (let i = 0; i < serializedChildPropertyNameParts.length - 1; ++i) {
-              const namePart: string = serializedChildPropertyNameParts[i];
-              if (!serializedPropertyParent[namePart]) {
-                break;
-              } else {
-                serializedPropertyParent = serializedPropertyParent[namePart];
+              const childPropertyPath: PropertyPath = propertyPath.concat([childPropertyName], serializedChildPropertyNameParts);
+
+              // Get the child property's value spec.
+              const childPropertyValueSpec: TypeSpec<any, any> = resolveValueSpec(options, childPropertyPath, childPropertySpec.valueSpec);
+
+              const serializedChildPropertyValue: any = (!serializedPropertyParent ? undefined : serializedPropertyParent[serializedChildPropertyName]);
+              if (serializedChildPropertyValue != undefined) {
+                result[childPropertyName] = childPropertyValueSpec.deserialize(childPropertyPath, serializedChildPropertyValue, options);
+              } else if (childPropertySpec.required && !childPropertySpec.constant) {
+                failDeserializeMissingRequiredPropertyCheck(options, childPropertyPath, childPropertyValueSpec);
               }
             }
+          }
+        }
+      }
+
+      return result;
+    }
+  };
+}
+
+function serializationDiscriminatorPropertyNameGetter(derivedTypes: PolymorphicDerivedTypes): string {
+  return derivedTypes.discriminatorPropertyName;
+}
 
-            if (serializedPropertyParent) {
-              serializedChildPropertyName = serializedChildPropertyNameParts[serializedChildPropertyNameParts.length - 1];
+function deserializationDiscriminatorPropertyNameGetter(derivedTypes: PolymorphicDerivedTypes): string {
+  return derivedTypes.discriminatorPropertySerializedName || derivedTypes.discriminatorPropertyName;
+}
+
+/**
+ * Get all of the CompositeTypeSpecs that the provided value must implement if the starting
+ * CompositeTypeSpec is the provided compositeSpec.
+ */
+function getAllPolymorphicCompositeTypeSpecs(value: CompositeType, propertyPath: PropertyPath, compositeSpec: CompositeTypeSpec, options: SerializationOptions, discriminatorPropertyNameGetter: (derivedTypes: PolymorphicDerivedTypes) => string): CompositeTypeSpec[] {
+  const result: CompositeTypeSpec[] = [];
+
+  let compositeSpecChanged = true;
+  while (compositeSpecChanged) {
```

Comment:
Is this a one time search while initializing the client or we do this search every time we serialize or deserialize a polymorphic composite type?
 IMO doing it once during initialization and then saving the result as a tree where every node contains an array of  itself and all its children would be a better approach. Then during [de]serialization we simply check if we can find an item in that array. If yes then we are good, otherwise just use the name of the node (i.e that type (which could be the uber base type or an intermediate base type) in the polymorphic tree).

Reply:
I created issue https://github.com/Azure/ms-rest-js/issues/64 for tracking this optimization work.

Reply:
As discussed, an even better approach would be to let the generator generate this information in the model spec as an array of model type names containing the name of that model and all of its children. Then at runtime we do a simple look up in that array. If we can find something then that is awesome. Then we [de]serialize as that child type otherwise we [de]serialize as that (parent) model type and log a warning that we could not find the provided model type.